import os
import glob

configfile: "config.yaml"

celllines = ['H1437', 'HCC1395', 'H2009', 'HCC1954']

#clair3_outdir  = expand("output/clair3/{cell_line}_{assembly}/", cell_line=list(cellline_map_cram.keys()), assembly=['chm13', 'grch38'])
#sniffles_files  = expand("output/sniffles/{cell_line}_{assembly}.vcf.gz", cell_line=list(cellline_map_cram.keys()), assembly=['chm13', 'grch38'])
#cutesv_files  = expand("output/cutesv/{cell_line}_{assembly}.vcf.gz", cell_line=list(cellline_map_cram.keys()), assembly=['chm13', 'grch38'])

def input_sra(wildcards):
    return glob.glob(f"{wildcards.cell_line}{wildcards.pair}.{wildcards.platform}/*/*")


rule sra2fastq:
    input:
        sra = input_sra
    output:
        fastq = "{cell_line}_{pair}_{platform}.fastq.gz"
    conda: "sra"
    threads: 10
    resources:
        mem_mb=16000,
        tmpdir="local_tmp/"
    shell:
        """
        fastq-dump --stdout {input.sra} | pigz -p {threads} - > {output.fastq}
        """

wildcard_constraints:
    cell_line = "[A-Za-z0-9]+",
    pair = "BL|T",
    assembly = "chm13|grch38",
    platform = "ont1"

module aligner_workflow:
    snakefile: "rules/minimap2.smk"
    config: config
use rule * from aligner_workflow as minimap2_workflow_*
include: "rules/sniffles2.smk"
include: "rules/savana.smk"
include: "rules/clair3.smk"

module severus_workflow:
    snakefile: "rules/severus.smk"
    config: config

use rule * from severus_workflow as severus_workflow_*
include: "rules/nanomonsv.smk"


all_fastq = expand("{cell_line}_{pair}_{platform}.fastq.gz", cell_line=celllines, pair=["T", "BL"], platform=['ont1'])
all_fastq_size = expand("{cell_line}_{pair}_{platform}.fastq.gz.sz", cell_line=celllines, pair=["T", "BL"], platform=['ont1'])

clair3_outdir  = expand(expand("output/clair3/{cell_line}_{platform}/{{pair}}/{{assembly}}/", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_files  = expand(expand("output/sniffles/{cell_line}_{platform}/{{pair}}/{{assembly}}.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_tbi_files  = expand(expand("output/sniffles/{cell_line}_{platform}/{{pair}}/{{assembly}}.vcf.gz.tbi", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_mosaic_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{pair}}/{{assembly}}.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_mosaic_tbi_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{pair}}/{{assembly}}.vcf.gz.tbi", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])

sniffles_multi_files  = expand(expand("output/sniffles/{cell_line}_{platform}/{{pair}}/{{assembly}}.snf", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
"output/sniffles/{cell_line}_{platform}/{assembly}_multi.vcf.gz"
sniffles_multi_vcf_files  = expand(expand("output/sniffles/{cell_line}_{platform}/{{assembly}}_multi.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

sniffles_mix  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{assembly}}_mixdown_mosaic.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
sniffles_mix_tbi  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{assembly}}_mixdown_mosaic.vcf.gz.tbi", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

sniffles_mo_multi_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{pair}}/{{assembly}}.snf", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_mo_multi_vcf_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{assembly}}_multi.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

nanomonsv_files  = expand(expand("output/nanomonsv/{cell_line}_{platform}/{{assembly}}_tnpair.txt", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
savana_outdir  = expand(expand("output/savana/{cell_line}_{platform}/{{assembly}}", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
severus_outdir  = expand(expand("output/severus/{cell_line}_{platform}/{{assembly}}", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
severus_outdir_repeat2  = expand(expand("output/severus/{cell_line}_{platform}/{{assembly}}_repeat2", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
severus_single_outdir  = expand(expand("output/severus_{platform}/{cell_line}_{{pair}}_{{assembly}}", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])

include: "rules/gafcall.smk"
include: "rules/gafcall_tnpair.smk"
include: "rules/gafcall_mixed.smk"
include: "rules/svision.smk"
include: "rules/seqtk.smk"

extract_linear_files  = expand(expand("output/align/{cell_line}_{{pair}}_{platform}_{{assembly}}l.def.gsv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
extract_graph_files  = expand(expand("output/align/{cell_line}_{{pair}}_{platform}_{{assembly}}g.def.gsv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])

svision_outdir  = expand(expand("output/svision/{cell_line}_{platform}/{{assembly}}", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

paf_mixed  = expand(expand("output/align/{cell_line}_{platform}/{{assembly}}_mixdown.paf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
gaf_mixed  = expand(expand("output/align/{cell_line}_{platform}/{{assembly}}g_mixdown.paf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

# new interface failed in one sample, increase memory cannot resolve the problem, separate the steps
#    output: output/minisv_pair/HCC1395_ont1_tumor_hg38l+tgs.rsv
#    log: .snakemake/slurm_logs/rule_minisv_tnpair_tgs_extract_tumor/7576263.log (check log file(s) for error details)
#    shell:
#
#
#        minisv.js e -n TUMOR -0b ~/data/pangenome_sv_benchmarking/minisv/data/hg38.cen-mask.bed output/align/HCC1395_T_ont1_grch38l.paf.gz output/align/HCC1395_T_ont1_chm13l.paf.gz output/align/HCC1395_T_ont1_chm13g.paf.gz ../severus_data_self/output/align/HCC1395_T_ont1_self.paf.gz | bash > output/minisv_pair/HCC1395_ont1_tumor_hg38l+tgs.rsv

minisv_pair = expand("output/minisv_pair/{cell_line}_{platform}_pair_hg38l+tgs.msv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])

minisv_pair_t2t = expand("output/minisv_pair/{cell_line}_{platform}_pair_chm13l_l+t+g+s.msv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])



minisv_mixed_mosaic_g = expand(expand("output/minisv_mosaic/{cell_line}_{platform}/{{assembly}}g_g+x_mosaic.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=['chm13', 'grch38'])
minisv_mixed_mosaic_l = expand(expand("output/minisv_mosaic/{cell_line}_{platform}/{{assembly}}l_l+x_mosaic.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=['chm13', 'grch38'])
minisv_mixed_mosaic_lg = expand(expand("output/minisv_mosaic/{cell_line}_{platform}/{{assembly}}l_l+g_mosaic.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=['chm13', 'grch38'])
minisv_mixed_mosaic_lt = expand("output/minisv_mosaic/{cell_line}_{platform}/grch38l_l+g_mosaic.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])
minisv_mixed_mosaic_ltg = expand("output/minisv_mosaic/{cell_line}_{platform}/grch38l_l+t+g_mosaic.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])

#minisv_stepwise_pair = expand("output/minisv_stepwise_pair/{cell_line}_{platform}_hg38l+tgs.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])

minisv_self_extract = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_self.rsv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"])

minisv_linear_def_extract = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l.def.rsv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'])

minisv_graph_def_extract = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}g.def.rsv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'])

minisv_single_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l_l+x_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])

minisv_single_graph_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l_l+g_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])

minisv_single_s_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l_l+s_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])

minisv_single_lg_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l_l+g_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])

minisv_single_ts_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_grch38l_t+s_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])

minisv_single_tg_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_grch38l_t+g_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])

minisv_single_tgs_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_grch38l_t+g+s_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])

tn_eval_length = expand("output/minisv_view/tn_{cell_line}_{platform}_eval_len.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])
single_eval_length = expand("output/minisv_view/single_{cell_line}_{platform}_eval_len.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])
mixed_eval_length = expand("output/minisv_view/mixed_{cell_line}_{platform}_merge_eval_len.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])


rule all:
    input:
        all_fastq, all_fastq_size,
        rules.minimap2_workflow_all.input,
        sniffles_files, sniffles_tbi_files,
        sniffles_mosaic_files, sniffles_mosaic_tbi_files,
        sniffles_multi_files, sniffles_multi_vcf_files,
        sniffles_mo_multi_files, sniffles_mo_multi_vcf_files,
        sniffles_mix, sniffles_mix_tbi,
        savana_outdir, clair3_outdir, 
        rules.severus_workflow_all.input,
        severus_outdir, 
        severus_single_outdir,
        severus_outdir_repeat2,
        nanomonsv_files,
        extract_linear_files, extract_graph_files,
        svision_outdir,
        minisv_pair, minisv_pair_t2t,
        minisv_self_extract,
        minisv_linear_def_extract, minisv_graph_def_extract, 
        minisv_single_call, minisv_single_graph_call, minisv_single_tg_call, minisv_single_tgs_call, minisv_single_lg_call,
        minisv_single_ts_call, minisv_single_s_call,
        paf_mixed, gaf_mixed, 
        minisv_mixed_mosaic_l, minisv_mixed_mosaic_g, minisv_mixed_mosaic_lg, minisv_mixed_mosaic_lt, minisv_mixed_mosaic_ltg,
        tn_eval_length, 
        single_eval_length, mixed_eval_length
    default_target: True 
