import os
import glob
import sys
import pandas as pd
from snakemake.utils import validate

configfile: "config.yaml"

clair3_outdir  = expand(expand("output/clair3/{cell_line}_{platform}/{{pair}}/{{assembly}}/", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_files  = expand(expand("output/sniffles/{cell_line}_{platform}/{{pair}}/{{assembly}}.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_tbi_files  = expand(expand("output/sniffles/{cell_line}_{platform}/{{pair}}/{{assembly}}.vcf.gz.tbi", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_mosaic_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{pair}}/{{assembly}}.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_mosaic_tbi_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{pair}}/{{assembly}}.vcf.gz.tbi", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])

sniffles_multi_files  = expand(expand("output/sniffles/{cell_line}_{platform}/{{pair}}/{{assembly}}.snf", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
"output/sniffles/{cell_line}_{platform}/{assembly}_multi.vcf.gz"
sniffles_multi_vcf_files  = expand(expand("output/sniffles/{cell_line}_{platform}/{{assembly}}_multi.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

sniffles_mo_multi_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{pair}}/{{assembly}}.snf", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_mo_multi_vcf_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{assembly}}_multi.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

nanomonsv_files  = expand(expand("output/nanomonsv/{cell_line}_{platform}/{{assembly}}_tnpair.txt", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
savana_outdir  = expand(expand("output/savana/{cell_line}_{platform}/{{assembly}}", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])


wildcard_constraints:
    cell_line = "[A-Za-z0-9]+",
    pair = "BL|T",
    assembly = "chm13|grch38",
    platform = "ont|hifi",
    rep = "1|2|3"

include: "rules/samtools.smk"
include: "rules/sniffles2.smk"

include: "rules/savana.smk"
include: "rules/clair3.smk"

module severus_workflow:
    snakefile: "rules/severus.smk"
    config: config

use rule * from severus_workflow as severus_workflow_*

include: "rules/nanomonsv.smk"
include: "rules/svision.smk"

rule minimap2_self:
    input:
        hap = expand("/homes6/alvin/data/pangenome_sv_benchmarking/5.assembly_evaluation/{{cell_line}}BL.asm.bp.hap{hap}.fa.gz", hap=[1,2]),
        fastq = os.path.join(config['prefix'], "{cell_line}{pair}.{platform}{rep}.fastq.gz")
    output:
        paf = "output/gafcall/{cell_line}{pair}_{platform}{rep}_self.paf.gz"
    threads: 24
    resources:
        mem_mb=64000
    shell:
        """
        ~/data/pangenome_sv_benchmarking/1a.alignment_sv_tools/minimap2/minimap2 --ds -t {threads} -cxlr:hq <(zcat {input.hap}) -I100g --secondary=no {input.fastq} | gzip - > {output.paf}
        """

rule minimap2_paf:
    input:
        fastq = os.path.join(config['prefix'], "{cell_line}{pair}.{platform}{rep}.fastq.gz"),
        assembly = "../1a.alignment_sv_tools/{assembly}.fa"
    output:
        paf = "output/gafcall/{cell_line}{pair}_{platform}{rep}_{assembly}.paf.gz"
    threads: 24
    shell:
        """
        ../1a.alignment_sv_tools/minimap2/minimap2 -cx lr:hq -t {threads} {input.assembly} {input.fastq} | gzip - > {output.paf}
        """

rule minigraph:
    input:
        fastq = os.path.join(config['prefix'], "{cell_line}{pair}.{platform}{rep}.fastq.gz"),
        assembly = "../1a.alignment_sv_tools/{assembly}.gfa.gz"
    output:
        gaf = "output/gafcall/{cell_line}{pair}_{platform}{rep}_{assembly}.gaf.gz"
    threads: 24
    shell:
        """
        ../1a.alignment_sv_tools/minigraph/minigraph -cxlr -t24 {input.assembly} {input.fastq} | gzip - > {output.gaf}
        """

bam_files = expand(expand("../1a.alignment_sv_tools/output/align/{cell_line}_{platform}/{{pair}}/{{assembly}}.cram", 
             cell_line = config['samples']['tumor'], platform=config['samples']['platform']),
             pair=['T', 'BL'], assembly=['chm13', 'grch38'])

self_files = expand(expand("output/gafcall/{cell_line}{{pair}}_{platform}{{rep}}_self.paf.gz", 
             cell_line = config['samples']['tumor'], platform=config['samples']['platform']),
             pair=['T', 'BL'], rep=[1,2], assembly=['chm13', 'grch38'])
paf_files = expand(expand("output/gafcall/{cell_line}{{pair}}_{platform}{{rep}}_{{assembly}}.paf.gz", 
             cell_line = config['samples']['tumor'], platform=config['samples']['platform']),
             pair=['T', 'BL'], rep=[1,2], assembly=['chm13', 'grch38'])
gaf_files = expand(expand("output/gafcall/{cell_line}{{pair}}_{platform}{{rep}}_{{assembly}}.gaf.gz", 
             cell_line = config['samples']['tumor'], platform=config['samples']['platform']),
             pair=['T', 'BL'], rep=[1,2], assembly=['chm13', 'grch38'])

severus_outdir  = expand(expand("output/severus/{cell_line}_{platform}/{{assembly}}", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
severus_single_outdir  = expand(expand("output/severus_{platform}/{cell_line}_{{pair}}_{{assembly}}", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])

svision_vcfs  = expand(expand("output/svision/{cell_line}_{platform}/{{assembly}}/{cell_line}.svision_pro_v1.8.s5.vcf", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

svision_vcfs_somatic  = expand(expand("output/svision/{cell_line}_{platform}/somatic/{cell_line}_{{assembly}}_tn.svision_pro_v1.8.s5.somatic_s1.vcf", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

rule all:
    input:
        bam_files, self_files, svision_vcfs_somatic,
        sniffles_files, sniffles_tbi_files,
        sniffles_mosaic_files, sniffles_mosaic_tbi_files,
        sniffles_multi_files, sniffles_multi_vcf_files,
        rules.severus_workflow_all.input,
        severus_outdir, severus_single_outdir,
        sniffles_mo_multi_files, sniffles_mo_multi_vcf_files,
        savana_outdir, clair3_outdir, 
        nanomonsv_files, paf_files, gaf_files, svision_vcfs
    default_target: True 
