import os
import glob
import sys
import pandas as pd
from snakemake.utils import validate

configfile: "config.yaml"

clair3_outdir  = expand(expand("output/clair3/{cell_line}_{platform}/{{pair}}/{{assembly}}/", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
clair3_mix10_outdir  = expand(expand("output/clair3_mix10/{cell_line}_{platform}/{{pair}}/{{assembly}}/", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
clair3_mix25_outdir  = expand(expand("output/clair3_mix25/{cell_line}_{platform}/{{pair}}/{{assembly}}/", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])

sniffles_files  = expand(expand("output/sniffles/{cell_line}_{platform}/{{pair}}/{{assembly}}.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_tbi_files  = expand(expand("output/sniffles/{cell_line}_{platform}/{{pair}}/{{assembly}}.vcf.gz.tbi", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_mosaic_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{pair}}/{{assembly}}.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_mosaic_tbi_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{pair}}/{{assembly}}.vcf.gz.tbi", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])

sniffles_mix_mosaic_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{assembly}}_mixdown_mosaic.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
sniffles_mix_mosaic_tbi_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{assembly}}_mixdown_mosaic.vcf.gz.tbi", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

sniffles_mix10_mosaic_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{assembly}}_mixdown10_mosaic.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
sniffles_mix10_mosaic_tbi_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{assembly}}_mixdown10_mosaic.vcf.gz.tbi", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

sniffles_multi_files  = expand(expand("output/sniffles/{cell_line}_{platform}/{{pair}}/{{assembly}}.snf", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
"output/sniffles/{cell_line}_{platform}/{assembly}_multi.vcf.gz"
sniffles_multi_vcf_files  = expand(expand("output/sniffles/{cell_line}_{platform}/{{assembly}}_multi.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

sniffles_mo_multi_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{pair}}/{{assembly}}.snf", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])
sniffles_mo_multi_vcf_files  = expand(expand("output/sniffles_mosaic/{cell_line}_{platform}/{{assembly}}_multi.vcf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

nanomonsv_files  = expand(expand("output/nanomonsv/{cell_line}_{platform}/{{assembly}}_tnpair.txt", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
savana_outdir  = expand(expand("output/savana/{cell_line}_{platform}/{{assembly}}", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
savana_vcf  = expand(expand("output/savana/{cell_line}_{platform}/{{assembly}}/{{assembly}}.classified.somatic.vcf", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

savana12_outdir  = expand(expand("output/savana12/{cell_line}_{platform}/{{assembly}}", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
savana12_vcf  = expand(expand("output/savana12/{cell_line}_{platform}/{{assembly}}/{{assembly}}_T_tag.classified.somatic.vcf", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

severus_outdir  = expand(expand("output/severus/{cell_line}_{platform}/{{assembly}}", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
severus_single_outdir  = expand(expand("output/severus_{platform}/{cell_line}_{{pair}}_{{assembly}}", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=config['assembly'])

severus_mixdown_outdir  = expand(expand("output/severus_{platform}/{cell_line}_{{assembly}}_mixdown", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

severus_mixdown_outdir_cutoff2  = expand(expand("output/severus_{platform}/{cell_line}_{{assembly}}_mixdown_cutoff2", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
severus_mixdown10_outdir_cutoff2  = expand(expand("output/severus_{platform}/{cell_line}_{{assembly}}_mixdown10_cutoff2", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

severus_mixdown25_outdir_cutoff2_withmixphased  = expand(expand("output/severus_{platform}/{cell_line}_{{assembly}}_mixdown25_cutoff2_phased_mixed", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
severus_mixdown10_outdir_cutoff2_withmixphased  = expand(expand("output/severus_{platform}/{cell_line}_{{assembly}}_mixdown10_cutoff2_phased_mixed", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])


# output/severus_{platform}_mixedsomatic/{cell_line}_{assembly}_mixdown25_cutoff2_phased_mixed_withBL
severus_mixdown25_outdir_cutoff2_withmixphased_somatic  = expand(expand("output/severus_{platform}_mixedsomatic/{cell_line}_{{assembly}}_mixdown25_cutoff2_phased_mixed_withBL", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
severus_mixdown10_outdir_cutoff2_withmixphased_somatic  = expand(expand("output/severus_{platform}_mixedsomatic/{cell_line}_{{assembly}}_mixdown10_cutoff2_phased_mixed_withBL", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

svision_outdir  = expand(expand("output/svision/{cell_line}_{platform}/{{assembly}}", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

##output/svision/HCC1954_hifi1/chm13/HCC1954_chm13_tn.svision_pro_v1.8.s5.somatic_s1.vcf
svision_outdir_somatic = expand(expand("output/svision/{cell_line}_{platform}/somatic/{cell_line}_{{assembly}}_tn.svision_pro_v1.8.s5.somatic_s1.vcf", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])

svision_single_vcfs  = expand(expand("output/svision_single/{cell_line}_{platform}_{{pair}}/{{assembly}}/{cell_line}.svision_pro_v1.8.s5.vcf", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'], pair=['T', 'BL'])

wildcard_constraints:
    cell_line = "[A-Za-z0-9]+",
    pair = "BL|T",
    assembly = "chm13|grch38",
    platform = "ont1|ont2|hifi1"

module aligner_workflow:
    snakefile: "rules/minimap2.smk"
    config: config

use rule * from aligner_workflow as minimap2_workflow_*

include: "rules/sniffles2.smk"
include: "rules/savana.smk"
include: "rules/savana1.2.smk"
include: "rules/clair3.smk"

module severus_workflow:
    snakefile: "rules/severus.smk"
    config: config

use rule * from severus_workflow as severus_workflow_*

module severus_workflow_cutoff:
    snakefile: "rules/severus_cutoffs.smk"
    config: config

use rule * from severus_workflow_cutoff as severus_workflow_cutoff_*

include: "rules/nanomonsv.smk"
include: "rules/svision.smk"
include: "rules/gafcall_tnpair.smk"
include: "rules/gafcall_mixed.smk"
include: "rules/hifiasm.smk"

#minisv_self_extract = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_self.rsv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"])
#
#minisv_linear_def_extract = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l.def.rsv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'])
#
#minisv_graph_def_extract = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}g.def.rsv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'])
#
#minisv_single_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l_l+x_merge_c5s0.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'])
#
#minisv_single_graph_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l_l+g_merge_c5s0.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'])
#
#minisv_single_s_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l_l+s_merge_c5s0.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'])
#
#
#minisv_single_ts_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_grch38l_t+s_merge_c5s0.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'])
#
#minisv_single_tg_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_grch38l_t+g_merge_c5s0.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'])
#
#minisv_single_tgs_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_grch38l_t+g+s_merge_c5s0.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'])

#paf_mixed  = expand(expand("output/align/{cell_line}_{platform}/{{assembly}}_mixdown.paf.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=config['assembly'])
#
#minisv_pair = expand("output/minisv_pair/{cell_line}_{platform}_pair_hg38l+tgs.msv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])
#
#minisv_pair_t2t = expand("output/minisv_pair/{cell_line}_{platform}_pair_chm13l_l+t+g+s.msv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])
#
#
#minisv_mixed_mosaic_g = expand(expand("output/minisv_mosaic/{cell_line}_{platform}/{{assembly}}g_g+x_mosaic.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=['chm13', 'grch38'])
#minisv_mixed_mosaic_l = expand(expand("output/minisv_mosaic/{cell_line}_{platform}/{{assembly}}l_l+x_mosaic.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=['chm13', 'grch38'])
#minisv_mixed_mosaic_lg = expand(expand("output/minisv_mosaic/{cell_line}_{platform}/{{assembly}}l_l+g_mosaic.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=['chm13', 'grch38'])
#minisv_mixed_mosaic_lt = expand("output/minisv_mosaic/{cell_line}_{platform}/grch38l_l+g_mosaic.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])
#minisv_mixed_mosaic_ltg = expand("output/minisv_mosaic/{cell_line}_{platform}/grch38l_l+t+g_mosaic.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])
#
#minisv_stepwise_pair = expand("output/minisv_stepwise_pair/{cell_line}_{platform}_hg38l+tgs.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])
#
#minisv_self_extract = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_self.rsv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"])
#
#minisv_linear_def_extract = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l.def.rsv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'])
#
#minisv_graph_def_extract = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}g.def.rsv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'])
#
#minisv_single_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l_l+x_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])
#
#minisv_single_graph_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l_l+g_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])
#
#minisv_single_s_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l_l+s_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])
#
#minisv_single_lg_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_{{assembly}}l_l+g_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])
#
#minisv_single_ts_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_grch38l_t+s_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])
#
#minisv_single_tg_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_grch38l_t+g_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])
#
#minisv_single_tgs_call = expand(expand("output/minisv/{cell_line}_{{pair}}_{platform}_grch38l_t+g+s_merge_{{cnt}}.msv.gz", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=["T", "BL"], assembly=['chm13', 'grch38'], cnt=['c2s0', 'c3s0', 'c4s0', 'c5s0'])
#
#tn_eval_length = expand("output/minisv_view/tn_{cell_line}_{platform}_eval_len.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])
#single_eval_length = expand("output/minisv_view/single_{cell_line}_{platform}_eval_len.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])
#mixed_eval_length = expand("output/minisv_view/mixed_{cell_line}_{platform}_merge_eval_len.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform'])


include: "rules/minisv_eval_all_tools.smk"

minisv_eval_files = expand(expand("output/minisv_eval_tnpair/{cell_line}_{platform}_{{assembly}}_count{{cutoff}}_eval.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=['chm13', 'grch38'], cutoff=[2,3,4,5,10])

minisv_eval_concensus = expand(expand("output/minisv_eval_tnpair_concensus/{cell_line}_{platform}_{{assembly}}_count{{cutoff}}_eval.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=['chm13', 'grch38'], cutoff=[2,3,4,5,10])

single_eval_performance = expand(expand("output/minisv_eval_single{{pair}}/single_{cell_line}_{platform}_{{assembly}}_count{{cutoff}}_eval.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=['T', 'BL'], assembly=['chm13', 'grch38'], cutoff=[2,3,4,5,10])

single_eval_performance_20kb = expand(expand("output/minisv_eval_single{{pair}}/single_{cell_line}_{platform}_{{assembly}}_count{{cutoff}}_eval_len20kb.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=['T', 'BL'], assembly=['chm13', 'grch38'], cutoff=[2,3,4,5,10])
single_eval_performance_20kb_concensus = expand(expand("output/minisv_eval_single{{pair}}/single_{cell_line}_{platform}_{{assembly}}_count{{cutoff}}_eval_len20kb_concensus.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=['T', 'BL'], assembly=['chm13', 'grch38'], cutoff=[2,3,4,5,10])

single_eval_performance_100kb = expand(expand("output/minisv_eval_single{{pair}}/single_{cell_line}_{platform}_{{assembly}}_count{{cutoff}}_eval_len100kb.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=['T', 'BL'], assembly=['chm13', 'grch38'], cutoff=[2,3,4,5,10])
single_eval_performance_100kb_concensus = expand(expand("output/minisv_eval_single{{pair}}/single_{cell_line}_{platform}_{{assembly}}_count{{cutoff}}_eval_len100kb_concensus.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=['T', 'BL'], assembly=['chm13', 'grch38'], cutoff=[2,3,4,5,10])

minisv_eval_mixed_files = expand(expand("output/minisv_eval_tnpair/{cell_line}_{platform}_{{assembly}}_count{{cutoff}}_eval_mixed.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=['chm13', 'grch38'], cutoff=[2,3,4,5,10])

tn_eval_length = expand(expand("output/alltools_view/tn_{cell_line}_{platform}_{{assembly}}_eval_len.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), assembly=['chm13', 'grch38'])

#
#pair_eval_single_length = expand(expand("output/minisv_view_single{{pair}}/single_{cell_line}_{platform}_{{assembly}}_eval_len.tsv", zip, cell_line=config['samples']['tumor'], platform=config['samples']['platform']), pair=['T', 'BL'], assembly=['chm13', 'grch38'])

# hifiasm only applies to hifi
cell_line_hifi = [i for i,j in zip(config['samples']['tumor'], config['samples']['platform']) if 'ont' not in j]
platform_hifi  = [j for i,j in zip(config['samples']['tumor'], config['samples']['platform']) if 'ont' not in j]

hifiasm_mixed_fastq = expand(expand("output/align/{cell_line}_{platform}/{{assembly}}_mixdown.fastq.gz", zip, cell_line = cell_line_hifi, platform = platform_hifi), assembly=['chm13', 'grch38'])
hifiasm_mixed_fastq10 = expand(expand("output/align/{cell_line}_{platform}/{{assembly}}_mixdown10.fastq.gz", zip, cell_line = cell_line_hifi, platform = platform_hifi), assembly=['chm13', 'grch38'])

#hifiasm_mixed_asm1 = expand("output/align/{cell_line}_{platform}/chm13_mixdown.asm.bp.hap1.p_ctg.gfa", zip, cell_line = cell_line_hifi, platform = platform_hifi)
#hifiasm_mixed_asm2 = expand("output/align/{cell_line}_{platform}/chm13_mixdown.asm.bp.hap2.p_ctg.gfa", zip, cell_line = cell_line_hifi, platform = platform_hifi)
hifiasm_mixed10_asm1 = expand("output/align/{cell_line}_{platform}/chm13_mixdown10.asm.bp.hap1.p_ctg.gfa", zip, cell_line = cell_line_hifi, platform = platform_hifi)
hifiasm_mixed10_asm2 = expand("output/align/{cell_line}_{platform}/chm13_mixdown10.asm.bp.hap2.p_ctg.gfa", zip, cell_line = cell_line_hifi, platform = platform_hifi)

haplotag_mixed10 = expand(expand(os.path.join(config['pwd'], "output/align/{cell_line}_{platform}_{{assembly}}_mixdown10_tag.cram"), zip, cell_line = cell_line_hifi, platform = platform_hifi), assembly=['chm13', 'grch38'])
haplotag_mixed25 = expand(expand(os.path.join(config['pwd'], "output/align/{cell_line}_{platform}_{{assembly}}_mixdown25_tag.cram"), zip, cell_line = cell_line_hifi, platform = platform_hifi), assembly=['chm13', 'grch38'])

snf_mixed10_withBL = expand(expand("output/sniffles/{cell_line}_{platform}/{{assembly}}_multi_mixdown10.vcf.gz", zip, cell_line = cell_line_hifi, platform = platform_hifi), assembly=['chm13', 'grch38'])

snf_mixed25_withBL = expand(expand("output/sniffles/{cell_line}_{platform}/{{assembly}}_multi_mixdown25.vcf.gz", zip, cell_line = cell_line_hifi, platform = platform_hifi), assembly=['chm13', 'grch38'])


rule all:
    input:
        rules.minimap2_workflow_all.input,
        rules.severus_workflow_all.input,
        rules.severus_workflow_cutoff_all.input,
        sniffles_files, sniffles_tbi_files,
        sniffles_mosaic_files, sniffles_mosaic_tbi_files,
        sniffles_multi_files, sniffles_multi_vcf_files,
        sniffles_mo_multi_files, sniffles_mo_multi_vcf_files,
        savana_outdir, savana12_outdir, savana_vcf, savana12_vcf,
        clair3_outdir, clair3_mix10_outdir, clair3_mix25_outdir,
        severus_outdir, severus_single_outdir,
        nanomonsv_files,
        svision_outdir, svision_single_vcfs,
        severus_mixdown_outdir, severus_mixdown_outdir_cutoff2, severus_mixdown10_outdir_cutoff2,
        severus_mixdown25_outdir_cutoff2_withmixphased_somatic, severus_mixdown10_outdir_cutoff2_withmixphased_somatic,

        svision_outdir_somatic,
        sniffles_mix_mosaic_files, sniffles_mix_mosaic_tbi_files,
        sniffles_mix10_mosaic_files, sniffles_mix10_mosaic_tbi_files, # 10% mixdown reads
        snf_mixed10_withBL, snf_mixed25_withBL,

        hifiasm_mixed_fastq10,
        haplotag_mixed10, haplotag_mixed25,
        severus_mixdown10_outdir_cutoff2_withmixphased, severus_mixdown25_outdir_cutoff2_withmixphased,
        #minisv_self_extract,
        #minisv_linear_def_extract, minisv_graph_def_extract, 
        #minisv_single_call, minisv_single_graph_call, minisv_single_tg_call, minisv_single_tgs_call,
        #minisv_single_ts_call, minisv_single_s_call,
        #paf_mixed
        tn_eval_length, 
        #single_eval_length, mixed_eval_length,  pair_eval_single_length,
        minisv_eval_files, minisv_eval_mixed_files, # tn_eval_length, 
        minisv_eval_concensus,
        single_eval_performance, single_eval_performance_20kb, single_eval_performance_20kb_concensus, single_eval_performance_100kb, single_eval_performance_100kb_concensus,
        "output/minisv_view_single/single_sample_allsvtools_svlength.tsv", 
        hifiasm_mixed_fastq, #hifiasm_mixed_asm1, hifiasm_mixed_asm2
        #hifiasm_mixed10_asm1, hifiasm_mixed10_asm2
    default_target: True 
